nanomq/pub_handler.c

```diff
@@ -238,129 +339,185 @@ static bool rule_engine_filter(nano_work *work, rule_engine_info *info)
-							} else {
-								cJSON_AddStringToObject(jso, "payload", payload);
-							}
-							break;
-						default:
-							break;
-						}
+			} else {
+			}
+			break;
+		case RULE_PAYLOAD_ALL:;
+			char *payload = pp->payload.data;
+			cJSON *jp = cJSON_ParseWithLength(payload, pp->payload.len);
+
+			if (info->as[j]) {
+				if (jp) {
+					cJSON_AddItemToObject(jso, info->as[j], jp);
+				} else {
+					cJSON_AddStringToObject(
+					    jso, info->as[j], payload);
+				}
+			} else {
+				if (jp) {
+					cJSON_AddItemToObject(jso, "payload", jp);
+				} else {
+					cJSON_AddStringToObject(
+					    jso, "payload", payload);
+				}
+			}
+			break;
+		case RULE_PAYLOAD_FIELD:
+			for (int pi = 0; pi < cvector_size(info->payload);
+			     pi++) {
+				if (info->payload[pi]->is_store) {
+					switch (info->payload[pi]->type) {
+					case cJSON_Number:
+						if (info->payload[pi]->pas)
+						cJSON_AddNumberToObject(jso,
+						    info->payload[pi]->pas,
+						    (int) info->payload[pi]->value);
+						break;
+					case cJSON_String:
+						cJSON_AddStringToObject(jso,
+						    info->payload[pi]->pas,
+						    (char *) info->payload[pi]->value);
+						break;
+					case cJSON_Object:
+						cJSON_AddItemToObject(jso,
+						    info->payload[pi]->pas,
+						    (cJSON*) info->payload[pi]->value);
+						break;
+					default:
+						break;
 					}
-
 				}
+			}
+			break;
 
-				char *dest = cJSON_PrintUnformatted(jso);
-				// puts(dest);
+		default:
+			break;
+		}
+	}
 
-				fdb_transaction_set(work->config->tran, pp->var_header.publish.topic_name.body, pp->var_header.publish.topic_name.len, dest, strlen(dest));
-				FDBFuture *f = fdb_transaction_commit(work->config->tran);
+	return 0;
+}
+
+static int
+rule_engine_insert_sql(nano_work *work)
+{
+	rule_engine_info  *rule_infos = work->config->rule_engine;
+	size_t             rule_size  = cvector_size(rule_infos);
+	pub_packet_struct *pp         = work->pub_packet;
+	conn_param        *cp         = work->cparam;
 
-				// struct ResultSet* rs = newResultSet();
-				// if(getError(fdb_future_block_until_ready(f), "GetManySequential (block for get)", rs)) return;
-				fdb_future_destroy(f);
-				cJSON_free(dest);
-				cJSON_Delete(jso);
+	for (size_t i = 0; i < rule_size; i++) {
+		if (rule_engine_filter(work, &rule_infos[i])) {
+			cJSON *jso = NULL;
+			jso        = cJSON_CreateObject();
+			for (size_t j = 0; j < 9; j++) {
+				add_info_to_json(&rule_infos[i], jso, j, work);
 			}
+
+			char *dest = cJSON_PrintUnformatted(jso);
+			puts(dest);
+			fdb_transaction_set(work->config->tran,
+			    pp->var_header.publish.topic_name.body,
+			    pp->var_header.publish.topic_name.len, dest,
+			    strlen(dest));
+			FDBFuture *f =
+			    fdb_transaction_commit(work->config->tran);
+
+			fdb_future_destroy(f);
+			cJSON_free(dest);
+			cJSON_Delete(jso);
+		}
 	}
 
 	return 0;
 }
 #endif
 
-
 reason_code
 handle_pub(nano_work *work, struct pipe_content *pipe_ct, uint8_t proto)
 {
```

Comment:
need to check if we use handle_pub only in RECV state

Reply:
That's been checked.